### ISSUE TITLE
mpv cannot use 5.1 surround with ALSA output

### ISSUE THREAD

**Issue by szaszg** [mpv-player/mpv#2162]
1. mpv open ALSA surround5.1 device successfully
+ say: 'Got supported channel map: stereo (type FIXED)' ???
+ cannot set 'ALSA channel map: FL FR'
+ after all uses '48000Hz stereo 2ch float' output ???

Of course 'speaker-test -D surround51 -c 6 -t wav' works well and 'aplay' can play 5.1 surround music with surround51 ALSA device...

---
[cplayer] Command line options: '-v' '-ao' 'alsa=device=surround51' '/home/gergely/6_Channel_ID.mkv' '--end' '0'
[cplayer] mpv 0.9.2-git-6beed16 (C) 2000-2015 mpv/MPlayer/mplayer2 projects
[cplayer]  built on 2015-07-19T20:07:52
[cplayer] ffmpeg library versions:
[cplayer]    libavutil       54.27.100
[cplayer]    libavcodec      56.41.100
[cplayer]    libavformat     56.36.100
[cplayer]    libswscale      3.1.101
[cplayer]    libavfilter     5.16.101
[cplayer]    libswresample   1.2.100
[cplayer] 
[cplayer] Configuration: ./waf configure --prefix=/home/gergely/bin/av --disable-sdl2 --disable-sdl1 --enable-cdda
[cplayer] List of enabled features: alsa asm atomics audio-input av-pix-fmt-mmal avcodec-chroma-pos-api avframe-metadata avframe-skip-samples build-date cdda cplayer debug-build dlopen drm dvbin dvdnav dvdread encoding fchmod gl gl-x11 glibc-thread-name glob iconv jpeg libass libass-osd libav libavdevice libavfilter libdl libm librt libswresample linux-fstatfs lua nanosleep optimize oss-audio oss-audio-native posix posix-or-mingw posix-spawn pthreads pvr resampler shm stdatomic subprocess termios tv tv-v4l2 videodev vt.h x11 xext xinerama xrandr xv zlib
[global] search dirs: /home/gergely/.mpv /home/gergely/bin/av/etc/mpv
[global] search dirs: /home/gergely/.mpv /home/gergely/bin/av/etc/mpv
[global] config file: 'mpv.conf|config'
[global]     -> '/home/gergely/.mpv/config'
[cplayer] Reading config file /home/gergely/.mpv/config
[cplayer] Setting option 'vo' = 'xv' (flags = 4)
[cplayer] Setting option 'fs' = 'yes' (flags = 4)
[cplayer] Setting option 'dvd-device' = '/dev/sr0' (flags = 4)
[cplayer] Setting option 'cdrom-device' = '/dev/sr0' (flags = 4)
[cplayer] Setting option 'softvol' = 'yes' (flags = 4)
[cplayer] Setting option 'gapless-audio' = 'yes' (flags = 4)
[cplayer] Setting option 'v' = '' (flags = 8)
[cplayer] Setting option 'ao' = 'alsa=device=surround51' (flags = 8)
[cplayer] Setting option 'end' = '0' (flags = 8)
[global] search dirs: /home/gergely/.mpv /home/gergely/bin/av/etc/mpv
[global] config file: 'input.conf'
[global]     -> '/home/gergely/.mpv/input.conf'
[global] user path: '/home/gergely/.mpv/input.conf' -> '/home/gergely/.mpv/input.conf'
[file] Opened: /home/gergely/.mpv/input.conf
[input] Parsing input config file /home/gergely/.mpv/input.conf
[input] Input config file /home/gergely/.mpv/input.conf parsed: 3 binds
[osc] Loading script @osc.lua...
[global] search dirs: /home/gergely/.mpv /home/gergely/bin/av/etc/mpv
[global] config file: 'scripts'
[global] search dirs: /home/gergely/.mpv /home/gergely/bin/av/etc/mpv
[global] config file: 'lua'
[osc] loading mp.defaults
[osc] loading @osc.lua
[global] search dirs: /home/gergely/.mpv /home/gergely/bin/av/etc/mpv
[global] config path: 'lua-settings/osc.conf' -> '(NULL)'
[osc] lua-settings/osc.conf not found. 
[cplayer] Done loading @osc.lua.
[ytdl_hook] Loading script @ytdl_hook.lua...
[global] search dirs: /home/gergely/.mpv /home/gergely/bin/av/etc/mpv
[global] config file: 'scripts'
[global] search dirs: /home/gergely/.mpv /home/gergely/bin/av/etc/mpv
[global] config file: 'lua'
[ytdl_hook] loading mp.defaults
[ytdl_hook] loading @ytdl_hook.lua
[cplayer] Run command: hook_add, flags=0, args=[on_load, 1, 10]
[cplayer] Done loading @ytdl_hook.lua.
[global] search dirs: /home/gergely/.mpv /home/gergely/bin/av/etc/mpv
[global] config file: 'scripts'
[global] search dirs: /home/gergely/.mpv /home/gergely/bin/av/etc/mpv
[global] config file: 'lua'
[global] search dirs: /home/gergely/.mpv /home/gergely/bin/av/etc/mpv
[global] config path: 'watch_later/2B8BB18BF0465764D76240F03AC01349' -> '(NULL)'
[global] search dirs: /home/gergely/.mpv /home/gergely/bin/av/etc/mpv
[global] config path: 'watch_later' -> '(NULL)'
[global] search dirs: /home/gergely/.mpv /home/gergely/bin/av/etc/mpv
[global] config path: 'watch_later/2B8BB18BF0465764D76240F03AC01349' -> '(NULL)'
[global] search dirs: /home/gergely/.mpv /home/gergely/bin/av/etc/mpv
[global] config path: 'watch_later' -> '(NULL)'
[cplayer] Playing: /home/gergely/6_Channel_ID.mkv
[cplayer] Running hook: ytdl_hook/on_load
[cplayer] Run command: hook_ack, flags=0, args=[on_load]
[file] Opened: /home/gergely/6_Channel_ID.mkv
[demux] Trying demuxer: disc (force-level: normal)
[demux] Trying demuxer: edl (force-level: normal)
[demux] Trying demuxer: cue (force-level: normal)
[demux] Trying demuxer: rawaudio (force-level: normal)
[demux] Trying demuxer: rawvideo (force-level: normal)
[demux] Trying demuxer: tv (force-level: normal)
[demux] Trying demuxer: libass (force-level: normal)
[demux] Trying demuxer: mkv (force-level: normal)
[mkv] Found the head...
[mkv] + a segment...
[mkv] /---- [ parsing seek head ] ---------
[mkv] \---- [ parsing seek head ] ---------
[mkv] |+ segment information...
[mkv] | + timecode scale: 1000000
[mkv] | + duration: 5.838s
[mkv] | + segment uid 14 80 df ad 2d ed c5 69 0b 37 4c 10 0a c0 2f a9
[mkv] |+ segment tracks...
[mkv] | + a track...
[mkv] |  + Track number: 1
[mkv] |  + Track type: Audio
[mkv] |  + Audio track
[mkv] |   + Sampling frequency: 44100.000000
[mkv] |   + Bit depth: 16
[mkv] |   + Channels: 6
[mkv] |  + Codec ID: A_VORBIS
[mkv] |  + CodecPrivate, length 7352
[mkv] |  + Language: und
[mkv] |+ found cluster
[mkv] All headers are parsed!
[demux] Detected file format: Matroska
[find_files] Loading external files in /home/gergely/
[global] search dirs: /home/gergely/.mpv /home/gergely/bin/av/etc/mpv
[global] config path: 'sub/' -> '(NULL)'
[cplayer]  (+) Audio --aid=1 (\*) (vorbis)
[ad] Codec list:
[ad]     lavc:vorbis - Vorbis
[ad]     lavc:libvorbis (vorbis) - libvorbis
[ad] Opening audio decoder lavc:vorbis
[ad] Requesting 1 threads for decoding.
[ad] Selected audio codec: Vorbis [lavc:vorbis]
[cplayer] Starting playback...
[af] Audio filter chain:
[af]   [in] 44100Hz 5.1 6ch floatp
[af]   [out] 44100Hz 5.1 6ch floatp
[af]   [ao] 44100Hz 5.1 6ch floatp
[ao] Trying preferred audio driver 'alsa'
[ao/alsa] Setting option 'device' = 'surround51' (flags = 0)
[ao/alsa] requested format: 44100 Hz, 5.1 channels, floatp
[ao/alsa] using ALSA version: 1.0.29
**[ao/alsa] opening device 'surround51'**
**[ao/alsa] Got supported channel map: stereo (type FIXED)**
[ao/alsa] trying to set ALSA channel map: FL FR
**[ao/alsa] Device does not support requested channel map (stereo)**
[ao/alsa] got buffersize=3072 samples
[ao/alsa] got period size 1024
[ao/alsa] device buffer: 3072 samples.
[ao/alsa] using soft-buffer of 9600 samples.
[cplayer] AO: [alsa] 48000Hz stereo 2ch float
[cplayer] AO: Description: ALSA audio output
[af] Adding filter lavrresample 
[af] Audio filter chain:
[af]   [in] 44100Hz 5.1 6ch floatp
[af]   [lavrresample] 48000Hz stereo 2ch float
[af]   [out] 48000Hz stereo 2ch float
**[af]   [ao] 48000Hz stereo 2ch float**
[cplayer] EOF code: 1  
[cplayer] 
[ad] Uninit audio filters...
[af] Removing filter lavrresample 
[ad] Uninit audio decoder.
[cplayer] 
[cplayer] Exiting... (Errors when loading file)
[ytdl_hook] Exiting...
[osc] Exiting...
[ao/alsa] draining...

**Comment by wm4** [comment]
>'-ao' 'alsa=device=surround51'

Do not do this. Then it should work. (Though you might need to select a hw device, since dmix can't do surround.)

**Comment by lachs0r** [comment]
Rather than using the surround51 device, you might be interested in this Wiki 
article:

https://github.com/mpv-player/mpv/wiki/ALSA:-Surround-Sound-and-Upmixing

**Comment by szaszg** [comment]
Thanks a lot: https://github.com/mpv-player/mpv/wiki/ALSA:-Surround-Sound-and-Upmixing makes the magic...

**Comment by rickyrockrat** [comment]
I know this is closed, and I know nobody really cares 6 years later, BUT. pulse audio is still painful to set up, and things still seem to work when it is removed. AND this is a page people will hit for a few more years.

The wiki links above are broken, as is the one below the 'Miscellaneous Technical Detail Articles' on the main wiki page.
That page is now:
https://github.com/mpv-player/mpv/wiki/ALSA-Surround-Sound-and-Upmixing

I want to play 5.1 sound. This should not take 2 days to figure out. I changed the motherboard, swapped the HDD, and MPV refuses to work.  Kernel is 4.20, and specially built for the Intel HEVC hardware decoding.

I am still having a similar issue with ALSA 1.1.8 and MPV 0.29.1. I tried all sorts of suggestions, but MPV insisted ALSA only had stereo channels.
I tried --audio-device=auto, --audio-channels=auto, I made sure my .asoundrc and /etc/asound.conf did not have any custom things.
for reference, I used --msg-level=all=v (and yes, I know I could have limited it to the modules I wanted, but I don't care)

I finally gave up. My solution is this (in addition to my -vo vdpau --hwdec=vdpau for hw accelerated decoding):

 --audio-channels=6 audio-device=alsa/mpvmap --alsa-ignore-chmap

where mpvmap in my asoundrc is ( and knowing my luck the mapping will change every movie):

pcm.mpvmap {
    type plug
    slave.pcm surround51
    slave.channels 6
    ttable {
       0.0 1 # FrontL->FrontL
       1.1 1 # FrontR->FrontR
       2.4 1 # RearL->Center
       3.5 4 # RearR->LFE 
       4.2 1 # Center->RearL
       5.3 1 # LFE -> RearR

    }
}
